File: backend/app/api/routes/vulnerabilities.py
from fastapi import APIRouter, Depends, HTTPException, Query, status
from sqlalchemy.orm import Session

from app.api.deps import get_db, require_current_user
from app.db.models import User
from app.schemas.ai import (
    VulnerabilityAIEnrichmentOut,
    VulnerabilityAIEnrichmentRunResponse,
    VulnerabilityAIEnrichmentTriggerRequest,
)
from app.schemas.vulnerability import (
    VulnerabilityCreate,
    VulnerabilityDetail,
    VulnerabilityListPage,
    VulnerabilityReadinessStats,
    VulnerabilityUpdate,
)
from app.services.vulnerability_catalog import (
    DuplicateVulnerabilityError,
    create_vulnerability,
    get_vulnerability,
    get_vulnerability_readiness_stats,
    list_vulnerabilities_page,
    update_vulnerability,
)
from app.services.auth import user_audit_details
from app.services.vulnerability_ai_enrichment import (
    BASIC_EXTRACTION_PROFILE_KEY,
    WEB_ENRICHMENT_PROFILE_KEY,
    enrich_vulnerability_auto,
    enrich_vulnerability_from_existing_data,
    enrich_vulnerability_with_web_search,
    list_vulnerability_ai_enrichments,
)
from app.workers.tasks import ai_enrich_vulnerability

router = APIRouter()


@router.get("", response_model=VulnerabilityListPage)
async def get_vulnerabilities(
    match_readiness: str | None = None,
    information_completeness: str | None = None,
    search: str | None = Query(default=None, max_length=256),
    severity_labels: str | None = Query(default=None, max_length=256),
    kev_status: bool | None = None,
    ai_enrichment_status: str | None = Query(default=None, max_length=64),
    offset: int = Query(default=0, ge=0),
    limit: int = Query(default=30, ge=1, le=300),
    db: Session = Depends(get_db),
) -> VulnerabilityListPage:
    try:
        return list_vulnerabilities_page(
            db,
            match_readiness=match_readiness,
            information_completeness=information_completeness,
            search=search,
            severity_labels=severity_labels,
            kev_status=kev_status,
            ai_enrichment_status=ai_enrichment_status,
            offset=offset,
            limit=limit,
        )
    except ValueError as exc:
        raise HTTPException(status_code=400, detail=str(exc)) from exc


@router.get("/readiness/stats", response_model=VulnerabilityReadinessStats)
async def get_vulnerability_readiness_statistics(
    db: Session = Depends(get_db),
) -> VulnerabilityReadinessStats:
    return get_vulnerability_readiness_stats(db)


@router.post("", response_model=VulnerabilityDetail, status_code=status.HTTP_201_CREATED)
async def post_vulnerability(
    payload: VulnerabilityCreate,
    current_user: User = Depends(require_current_user),
    db: Session = Depends(get_db),
) -> VulnerabilityDetail:
    try:
        return create_vulnerability(
            db,
            payload,
            actor_id=current_user.id,
            actor_details=user_audit_details(current_user),
        )
    except DuplicateVulnerabilityError as exc:
        raise HTTPException(status_code=409, detail=str(exc)) from exc
    except ValueError as exc:
        raise HTTPException(status_code=400, detail=str(exc)) from exc


@router.get("/{vulnerability_id}", response_model=VulnerabilityDetail)
async def get_vulnerability_detail(
    vulnerability_id: str,
    db: Session = Depends(get_db),
) -> VulnerabilityDetail:
    vulnerability = get_vulnerability(db, vulnerability_id)
    if vulnerability is None:
        raise HTTPException(status_code=404, detail="Vulnerability not found")
    return vulnerability


@router.get(
    "/{vulnerability_id}/ai-enrichments",
    response_model=list[VulnerabilityAIEnrichmentOut],
)
async def get_vulnerability_ai_enrichments(
    vulnerability_id: str,
    db: Session = Depends(get_db),
) -> list[VulnerabilityAIEnrichmentOut]:
    try:
        return list_vulnerability_ai_enrichments(db, vulnerability_id)
    except ValueError as exc:
        raise _ai_enrichment_http_error(exc) from exc


@router.post(
    "/{vulnerability_id}/ai-enrichments",
    response_model=VulnerabilityAIEnrichmentRunResponse,
)
async def post_vulnerability_ai_enrichment(
    vulnerability_id: str,
    payload: VulnerabilityAIEnrichmentTriggerRequest,
    current_user: User = Depends(require_current_user),
    db: Session = Depends(get_db),
) -> VulnerabilityAIEnrichmentRunResponse:
    profile_key = payload.profile_key or (
        WEB_ENRICHMENT_PROFILE_KEY
        if payload.layer == "web_enrichment"
        else BASIC_EXTRACTION_PROFILE_KEY
    )
    if payload.async_mode:
        try:
            task = ai_enrich_vulnerability.delay(
                vulnerability_id,
                payload.layer,
                profile_key,
                payload.allow_web_enrichment,
                payload.force_refresh,
            )
            return VulnerabilityAIEnrichmentRunResponse(
                async_queued=True,
                task_id=task.id,
            )
        except Exception:
            pass

    try:
        if payload.layer == "web_enrichment":
            enrichment = enrich_vulnerability_with_web_search(
                db,
                vulnerability_id,
                profile_key=profile_key,
                actor_id=current_user.id,
                actor_details=user_audit_details(current_user),
                force_refresh=payload.force_refresh,
            )
        elif payload.layer == "auto":
            enrichment = enrich_vulnerability_auto(
                db,
                vulnerability_id,
                allow_web_enrichment=payload.allow_web_enrichment,
                actor_id=current_user.id,
                actor_details=user_audit_details(current_user),
                force_refresh=payload.force_refresh,
            )
        else:
            enrichment = enrich_vulnerability_from_existing_data(
                db,
                vulnerability_id,
                profile_key=profile_key,
                actor_id=current_user.id,
                actor_details=user_audit_details(current_user),
                force_refresh=payload.force_refresh,
            )
    except ValueError as exc:
        raise _ai_enrichment_http_error(exc) from exc
    return VulnerabilityAIEnrichmentRunResponse(enrichment=enrichment)


@router.patch("/{vulnerability_id}", response_model=VulnerabilityDetail)
async def patch_vulnerability_detail(
    vulnerability_id: str,
    payload: VulnerabilityUpdate,
    current_user: User = Depends(require_current_user),
    db: Session = Depends(get_db),
) -> VulnerabilityDetail:
    try:
        vulnerability = update_vulnerability(
            db,
            vulnerability_id,
            payload,
            actor_id=current_user.id,
            actor_details=user_audit_details(current_user),
        )
    except ValueError as exc:
        raise HTTPException(status_code=400, detail=str(exc)) from exc
    if vulnerability is None:
        raise HTTPException(status_code=404, detail="Vulnerability not found")
    return vulnerability


def _ai_enrichment_http_error(exc: ValueError) -> HTTPException:
    detail = str(exc)
    if "Vulnerability not found" in detail:
        return HTTPException(status_code=404, detail=detail)
    return HTTPException(status_code=400, detail=detail)

File: frontend/src/api/vulnerabilities.ts
import { request } from "@/api/client";
import type {
  VulnerabilityAIEnrichmentStatus,
  VulnerabilityAIEnrichmentAcceptRequest,
  VulnerabilityAIEnrichmentAcceptResponse,
  VulnerabilityAIEnrichmentBatchRequest,
  VulnerabilityAIEnrichmentBatchDetail,
  VulnerabilityAIEnrichmentBatchResponse,
  VulnerabilityAIEnrichment,
  VulnerabilityAIEnrichmentRejectRequest,
  VulnerabilityAIEnrichmentRunResponse,
  VulnerabilityAIEnrichmentTriggerRequest,
  VulnerabilityCreate,
  VulnerabilityDetail,
  VulnerabilityInformationCompleteness,
  VulnerabilityListPage,
  VulnerabilityMatchReadiness,
  VulnerabilityReadinessStats,
  VulnerabilityReviewDetail,
  VulnerabilityReviewListPage,
  VulnerabilityReviewQueue,
  VulnerabilityReviewResolution,
  VulnerabilityReviewResolutionCreate,
  VulnerabilityUpdate
} from "@/api/types";

export interface VulnerabilityListParams {
  match_readiness?: VulnerabilityMatchReadiness;
  information_completeness?: VulnerabilityInformationCompleteness;
  search?: string;
  severity_labels?: string;
  kev_status?: boolean;
  ai_enrichment_status?: VulnerabilityAIEnrichmentStatus;
  offset?: number;
  limit?: number;
}

export function getVulnerabilities(params: VulnerabilityListParams = {}) {
  return request<VulnerabilityListPage>("/api/v1/vulnerabilities", {
    query: {
      match_readiness: params.match_readiness,
      information_completeness: params.information_completeness,
      search: params.search,
      severity_labels: params.severity_labels,
      kev_status: params.kev_status,
      ai_enrichment_status: params.ai_enrichment_status,
      offset: params.offset,
      limit: params.limit
    }
  });
}

export function getVulnerabilityReadinessStats() {
  return request<VulnerabilityReadinessStats>("/api/v1/vulnerabilities/readiness/stats");
}

export function getVulnerabilityReviews(
  queue: VulnerabilityReviewQueue = "open",
  limit = 60,
  offset = 0
) {
  return request<VulnerabilityReviewListPage>("/api/v1/vulnerability-reviews", {
    query: { queue, limit, offset }
  });
}

export function getVulnerabilityReview(vulnerabilityId: string) {
  return request<VulnerabilityReviewDetail>(`/api/v1/vulnerability-reviews/${vulnerabilityId}`);
}

export function createVulnerabilityReviewResolution(
  vulnerabilityId: string,
  body: VulnerabilityReviewResolutionCreate
) {
  return request<VulnerabilityReviewResolution>(
    `/api/v1/vulnerability-reviews/${vulnerabilityId}/resolutions`,
    { method: "POST", body }
  );
}

export function getVulnerability(vulnerabilityId: string) {
  return request<VulnerabilityDetail>(`/api/v1/vulnerabilities/${vulnerabilityId}`);
}

export function createVulnerability(body: VulnerabilityCreate) {
  return request<VulnerabilityDetail>("/api/v1/vulnerabilities", {
    method: "POST",
    body
  });
}

export function updateVulnerability(
  vulnerabilityId: string,
  body: VulnerabilityUpdate
) {
  return request<VulnerabilityDetail>(`/api/v1/vulnerabilities/${vulnerabilityId}`, {
    method: "PATCH",
    body
  });
}

export function getVulnerabilityAIEnrichments(vulnerabilityId: string) {
  return request<VulnerabilityAIEnrichment[]>(
    `/api/v1/vulnerabilities/${vulnerabilityId}/ai-enrichments`
  );
}

export function triggerVulnerabilityAIEnrichment(
  vulnerabilityId: string,
  body: VulnerabilityAIEnrichmentTriggerRequest = {
    layer: "existing_data_extraction",
    async_mode: false
  }
) {
  return request<VulnerabilityAIEnrichmentRunResponse>(
    `/api/v1/vulnerabilities/${vulnerabilityId}/ai-enrichments`,
    {
      method: "POST",
      body
    }
  );
}

export function acceptVulnerabilityAIEnrichment(
  enrichmentId: string,
  body: VulnerabilityAIEnrichmentAcceptRequest
) {
  return request<VulnerabilityAIEnrichmentAcceptResponse>(
    `/api/v1/vulnerability-ai-enrichments/${enrichmentId}/accept`,
    {
      method: "POST",
      body
    }
  );
}

export function rejectVulnerabilityAIEnrichment(
  enrichmentId: string,
  body: VulnerabilityAIEnrichmentRejectRequest
) {
  return request<VulnerabilityAIEnrichment>(
    `/api/v1/vulnerability-ai-enrichments/${enrichmentId}/reject`,
    {
      method: "POST",
      body
    }
  );
}

export function createVulnerabilityAIEnrichmentBatch(
  body: VulnerabilityAIEnrichmentBatchRequest
) {
  return request<VulnerabilityAIEnrichmentBatchResponse>(
    "/api/v1/vulnerability-ai-enrichments/batch",
    {
      method: "POST",
      body
    }
  );
}

export function getVulnerabilityAIEnrichmentBatchDetail(batchRunId: string) {
  return request<VulnerabilityAIEnrichmentBatchDetail>(
    `/api/v1/vulnerability-ai-enrichments/batch/${batchRunId}`
  );
}
